> The first example is already tested.

Is it? Currently I get:
```pycon
>>> List[TypeVar][int]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/serhiy/py/cpython/Lib/typing.py", line 360, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/serhiy/py/cpython/Lib/typing.py", line 1400, in __getitem__
    new_args = self._determine_new_args(args)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/serhiy/py/cpython/Lib/typing.py", line 1437, in _determine_new_args
    new_arg = substfunc(new_arg_by_param[old_arg])
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: TypeVar.__typing_subst__() missing 1 required positional argument: 'arg'
```
It should raise the same error as for `List[int][int]`: `TypeError: typing.List[int] is not a generic class`.

> The second example won't trigger this code path.

Yes, but it means that the C code needs a similar fix.

> The fourth example won't trigger this code path either as unions with `TypeVar` automatically become `typing.Union` (which is a `typing._GenericAlias`).

There is no `TypeVar` nor `typing.Union` in that example.

I propose to add the following test:
```python
for a in List, list:
    for b in int, TypeVar, TypeVarTuple, ParamSpec, types.GenericAlias, types.UnionType:
        with self.assertRaisesRegex(TypeError, '.* is not a generic class')
            a[b][str]
```
It tests both Python and C implementation of var substitution with all public classes which have `__typing_substr__` or `__parameters__` (and added a plain type for comparison).

--- a/Lib/test/test_typing.py
+++ b/Lib/test/test_typing.py
@@ -3668,15 +3668,16 @@ class GenericTests(BaseTestCase):
         class A:
             __parameters__ = (T,)
         # Bare classes should be skipped
-        for a in List, list:
+        for a in (List, list):
             for b in (A, int, TypeVar, TypeVarTuple, ParamSpec, types.GenericAlias, types.UnionType):
                 with self.subTest(generic=a, sub=b):
                     with self.assertRaisesRegex(TypeError, '.* is not a generic class'):
                         a[b][str]
         # Duck-typing anything that looks like it has __parameters__.
-        # This test is optional and failure is okay.
+        # These tests are optional and failure is okay.
         self.assertEqual(List[A()].__parameters__, (T,))
-
+        # C version of GenericAlias
+        self.assertEqual(list[A()].__parameters__, (T,))
 
 class ClassVarTests(BaseTestCase):
 

--- a/Lib/test/test_typing.py
+++ b/Lib/test/test_typing.py
@@ -3668,7 +3668,11 @@ class GenericTests(BaseTestCase):
         class A:
             __parameters__ = (T,)
         # Bare classes should be skipped
-        self.assertEqual(List[A].__parameters__, ())
+        for a in List, list:
+            for b in (A, int, TypeVar, TypeVarTuple, ParamSpec, types.GenericAlias, types.UnionType):
+                with self.subTest(generic=a, sub=b):
+                    with self.assertRaisesRegex(TypeError, '.* is not a generic class'):
+                        a[b][str]
         # Duck-typing anything that looks like it has __parameters__.
         # This test is optional and failure is okay.
         self.assertEqual(List[A()].__parameters__, (T,))

--- a/Misc/NEWS.d/next/Library/2022-07-06-16-01-08.gh-issue-94607.Q6RYfz.rst
+++ b/Misc/NEWS.d/next/Library/2022-07-06-16-01-08.gh-issue-94607.Q6RYfz.rst
@@ -1,1 +1,2 @@
 Fix subclassing complex generics with type variables in :mod:`typing`. Previously an error message saying ``Some type variables ... are not listed in Generic[...]`` was shown.
+:mod:`typing` no longer populates ``__parameters__`` with the ``__parameters__`` of a Python class.

--- a/Objects/genericaliasobject.c
+++ b/Objects/genericaliasobject.c
@@ -219,6 +219,10 @@ _Py_make_parameters(PyObject *args)
     for (Py_ssize_t iarg = 0; iarg < nargs; iarg++) {
         PyObject *t = PyTuple_GET_ITEM(args, iarg);
         PyObject *subst;
+        // A bare Python class isn't generic.
+        if (PyType_Check(t)) {
+            continue;
+        }
         if (_PyObject_LookupAttr(t, &_Py_ID(__typing_subst__), &subst) < 0) {
             Py_DECREF(parameters);
             return NULL;

